dangerzone/container_utils.py

```diff
@@ -175,27 +189,110 @@ def load_image_tarball() -> None:
+    """Get an image ID from a digest.
+    The sha256: prefix should be omitted from the digest.
+    """
+    runtime = Runtime()
+    # There is a "digest" filter that you can use with
+    # "podman images -f digest:<digest>", but it's only available
+    # for podman >=4.4 (and bookworm ships 4.3)
+    # So, fallback on the json format instead
+    cmd = [
+        str(runtime.path),
+        "images",
+        "--format",
+        "json",
+    ]
+    log.debug(" ".join(cmd))
+    process = subprocess_run(cmd, check=True, capture_output=True)
+
+    images = json.loads(process.stdout.decode().strip())  # type:ignore[attr-defined]
+    filtered_images = [
+        image["Id"] for image in images if image["Digest"] == f"sha256:{digest}"
+    ]
+
+    if not filtered_images:
+        raise errors.ImageNotPresentException(
+            f"Unable to find an image with digest {digest}"
+        )
+    return filtered_images[0]
+
+
+def expected_image_name() -> str:
+    image_name_path = get_resource_path("image-name.txt")
+    return image_name_path.read_text().strip("\n")
+
+
+def container_pull(
+    image: str, manifest_digest: str, callback: Optional[Callable] = None
+) -> None:
+    """Pull a container image from a registry."""
+    runtime = Runtime()
+    cmd = [str(runtime.path), "pull", f"{image}@sha256:{manifest_digest}"]
+    process = subprocess.Popen(
+        cmd,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        text=True,
+        bufsize=1,
+    )
+
+    if callback:
+        for line in process.stdout:  # type: ignore
+            callback(line)
+
+    process.wait()
+    if process.returncode != 0:
+        raise errors.ContainerPullException(
+            f"Could not pull the container image: {process.returncode}"
         )
-        add_image_tag(bad_tag, good_tag)
-        delete_image_tag(bad_tag)
 
-    log.info("Successfully installed container image")
+
+def get_local_image_digest(image: Optional[str] = None) -> str:
+    """
+    Returns a image hash from a local image name
+    """
+    expected_image = image or expected_image_name()
+    # Get the image hash from the "podman images" command.
+    # It's not possible to use "podman inspect" here as it
+    # returns the digest of the architecture-bound image
+    runtime = Runtime()
+    cmd = [str(runtime.path), "images", expected_image, "--format", "{{.Digest}}"]
+    log.debug(" ".join(cmd))
+    try:
+        result = subprocess_run(
+            cmd,
+            capture_output=True,
+            check=True,
+        )
+        output = result.stdout.decode().strip().split("\n")  # type:ignore[attr-defined]
+        # In some cases, the output can be multiple lines with the same digest
+        # sets are used to reduce them.
+        lines = set(output)
+        if len(lines) != 1:
+            raise errors.MultipleImagesFoundException(
+                f"Expected a single line of output, got {len(lines)} lines: {lines}"
+            )
```

Comment:
I think we'll trigger this case if there are no images installed as well:

```
$ podman images doesnotexist --format '{{.Digest}}'; echo $?
0
```

In that case, we should raise an `ImageNotPresentException`.